Bunch of things

vespakoen opened the pull request:
- Add live reloading for ExtendScript code
- yarn -> npm
- add debug start script that will show verbose logs
- use `"noLib": true` for ExtendScript's `tsconfig.json` and require `extendscript-es5-shim-ts` (should probably see if the babel extendscript transform is still necessary)
- assume the environment is CEP
- add bridge to communicate from ExtendScript to CEP
- add `console.log` support in ExtendScript (which prints to the CEP console)
- assume the environment is always CEP / get rid of all the isCEPEnvironment

fusepilot commented on src/js/extendscript.tsx:
```diff
@@ -1,34 +1,14 @@
+
+// @ts-ignore
+const Buffer = cep_node.Buffer
+const jsxBundle = readFileSync('./dist/index.js').toString()
```
This breaks the `build` task for me. The `./dist/index.js` file doesn't exist unless I run the `start` task before hand. This was part of the reason why I create the `manifest.json` file to confidently grab the transpiled/hashed parcel bundle names and call them at runtime. Like so:

````js
const manifest = fs.readJsonSync(path.join(extensionPath, "manifest.json"));
loadExtendscript(manifest["index.jsx.ts"]);
````

 I'm a little hesitant to remove this feature.

vespakoen replied:
Since we are bundling the ExtendScript file seperately (as the entry file) it will not get a strange suffix, so we can expect it to be named "index.js" in the dist folder.

Also, to get live reloading capabilities, we have to know the path on forehand, since `fs.readJsonSync('some/path')` will be transformed by parcel (at compile time) into a `Buffer(withTheContentsOfTheFile)`, it also works with path.join() but only with variables it can figure out at compile time, so not the manifest file (as it's not there yet)

vespakoen replied:
Also, if you want the manifest for other purposes, you can use this plugin: https://www.npmjs.com/package/parcel-plugin-bundle-manifest

fusepilot replied:
I'm cool with swapping my built-in manifest implementation out for `parcel-plugin-bundle-manifest` and also maybe removing it from the starter template. But before we decide that, we need to address the error that happens when running `build`. This is the error I get:

````
extendscript.tsx:7:34: ENOENT: nosuch file or directory, open '~/Workspace/parcel-plugin-cep-starter/dist/index.js'
   5 | // @ts-ignore
   6 | var Buffer = cep_node.Buffer;
>  7 | var jsxBundle = fs_1.readFileSync('./dist/index.js').toString();
     |                                  ^
   8 | cep_interface_1.evalExtendscript(jsxBundle);
   9 | var host = cep_interface_1.getHostEnvironment();
````

Are you not seeing the same?

vespakoen replied:
I am not seeing the same, but I just pushed a change that will first run the "clean" and "build-jsx" tasks before starting the watch processes (start-jsx & start-js).

Hope that makes it work for you =)

fusepilot replied:
Something doesn't feel right with requiring something in `dist` from `src`. I think we want to keep the manifest way of doing this.

vespakoen replied:
Unfortunately we cannot use the manifest file I think since we depend on the bundler to insert the contents of the file at compile time, it works like `brfs` (from browserify) if you are familiar with that: https://github.com/browserify/brfs

Without this we cannot have live reload for .jsx files, which is worth the hack I think.

I am not aware of another way of doing this.

fusepilot replied:
What if we used https://github.com/kentcdodds/babel-plugin-preval to find the name of the file at compile time?

vespakoen replied:
I think adding preval is also a hack, adds another dependency and might be more confusing in the end. My idea about using a process.env variable to get the path is also not going to work since we run 2 different bundlers, 1 for the JS and one for the JSX code, the JS bundler needs to find out the path to the JSX file but that bundler is not aware of the other one, I think we should keep it as it is and add a warning in the readme that when people rename / move the entry file of the extendscript code they also have to update the path in the JS code.

fusepilot commented on src/jsx/index.ts:
```diff
@@ -0,0 +1,29 @@
+/// <reference path="../../node_modules/types-for-adobe/shared/PlugPlugExternalObject.ts" />
+import 'extendscript-es5-shim-ts'
```
```suggestion
```

Let's put this in `tsconfig.json`

vespakoen replied:
The 'extendscript-es5-shim-ts' is not only type definitions, it also contains code (including JSON.stringify + JSON.parse), we have to import it in the code to also get the code loaded in

vespakoen replied:
There are no type definitions for JSON (since we are using `"noLib": true` now when bundling ExtendScript (for better correctness), the `extendscript-es5-shim-ts` package provides the types (and shim) for JSON and some others, I am not sure if we should still use the babel preset, I still have to look into that but at least they don't seem to conflict with eachother

fusepilot replied:
The babel preset is necessary if the user of the starter doesn't want to use Typescript.

I wonder if there should be 2 starter templates, one with Typescript ( and without babel ), and one with regular Javascript ( with babel and the preset )?

Also, Babel 7 can transpile Typescript now, so maybe we stick with babel and remove the typescript dependency?

I'll need to think more about it, but curious to know your thoughts.

fusepilot replied:
I'll keep this in for now. But I think it'd be good to eventually remove this requirement.

fusepilot commented on src/jsx/tsconfig.json:
```diff
@@ -4,6 +4,7 @@
     "moduleResolution": "node",
     "strict": true,
+    "noLib": true,
     "types": ["AfterEffects/2018"],
```
```suggestion
     "types": ["AfterEffects/2018", "extendscript-es5-shim-ts"],
```

vespakoen replied:
See above

fusepilot replied:
Will keep for now.

fusepilot commented on src/jsx/index.ts:
```diff
@@ -0,0 +1,29 @@
+/// <reference path="../../node_modules/types-for-adobe/shared/PlugPlugExternalObject.ts" />
```
```suggestion
```

Let's put this in `console.ts` where its immediately used.

vespakoen replied:
I agree

fusepilot commented on src/jsx/console.ts:
```diff
@@ -0,0 +1,35 @@
+import dispatch from './bridge'
```
```suggestion
/// <reference path="../../node_modules/types-for-adobe/shared/PlugPlugExternalObject.ts" />

import dispatch from './bridge'
```

vespakoen replied:
I will add it to bridge.ts instead

fusepilot commented on package.json:
```diff
@@ -9,34 +9,46 @@
+    "debug": "npm-run-all --parallel debug-jsx debug-js",
+    "debug-js": "parcel --log-level 4 src/js/index.html",
+    "debug-jsx": "parcel watch src/jsx/index.ts --no-hmr --no-source-maps --log-level 4",
+    "build": "npm-run-all clean build-js build-jsx",
```
```suggestion
    "build": "npm-run-all --serial clean build-jsx build-js",
```

This seems to fix the error I was experiencing with it not finding `./dist/index.js` and causing the build to break.

fusepilot commented on src/js/components/AdobeAppInfo.tsx:
```diff
@@ -1,47 +1,30 @@
 
-  render() {
+export default class AdobeAppInfo extends React.Component {
+  async render() {
```
Looks like this was resolved.

vespakoen commented:
I just pushed changes that runs the `build` command serially and starts a typescript server just for showing the typescript errors (not for actually compiling stuff)

I also fixed most .ts errors (I have only one left that comes from cep-interface):
```
../../node_modules/cep-interface/lib/cs-interface.d.ts:1:22 - error TS6053: File '/Users/koen/Projects/adobe-extension-tools/parcel-plugin-cep-starter/node_modules/cep-interface/src/window.d.ts' not found.

1 /// <reference path="../src/window.d.ts" />
```

And updated the parcel-plugin-cep dependency to version 1.2.0
